src/dtx/dtx_common.c

```diff
@@ -0,0 +1,632 @@
+		leader ? "leader" : "non-leader");
+
+	return 0;
+}
+
+int
+dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
+	struct ds_cont *cont, int result)
+{
+	int	rc = 0;
+
+	if (result < 0) {
+		if (!dth->dth_leader && dth->dth_dti_cos_count > 0) {
+			/* XXX: For non-leader replica, even if we fail to
+			 *	make related modification for some reason,
+			 *	we still need to commit the DTXs for CoS.
+			 *	Because other replica may have already
+			 *	committed them. For leader case, it is
+			 *	not important even if we miss to commit
+			 *	the CoS DTXs, because they are still in
+			 *	CoS cache, and can be committed next time.
+			 */
+			rc = vos_dtx_commit(cont->sc_hdl, dth->dth_dti_cos,
+					    dth->dth_dti_cos_count);
+			if (rc != 0)
+				D_ERROR(DF_UUID": Fail to DTX CoS commit: %d\n",
+					DP_UUID(cont->sc_uuid), rc);
+		}
+
+		goto fail;
+	}
+
+	if (!dth->dth_leader || UMMID_IS_NULL(dth->dth_ent) || dth->dth_non_rep)
+		goto out;
+
+	/* If the DTX is started befoe DTX resync operation (for rebuild),
+	 * then it is possbile that the DTX resync ULT may have aborted
+	 * current DTX before remote replica(s) modification by race. So
+	 * let's check DTX status locally before marking as 'committable'.
+	 */
+	if (dth->dth_handled_time <= cont->sc_dtx_resync_time) {
+		rc = vos_dtx_check_committable(cont->sc_hdl, NULL,
+					       &dth->dth_xid, 0, false);
+		if (rc < 0) {
+			result = (rc == -DER_NONEXIST ? -DER_INPROGRESS : rc);
+			D_GOTO(fail, result);
+		}
+	}
+
+	rc = vos_dtx_add_cos(dth->dth_coh, &dth->dth_oid, &dth->dth_xid,
+			     dth->dth_dkey_hash, dth->dth_handled_time,
+			     dth->dth_intent == DAOS_INTENT_PUNCH ?
+			     true : false);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": Fail to add DTX "DF_DTI" to CoS cache: %d. "
+			"Try to commit it sychronously.\n",
+			DP_UUID(cont->sc_uuid), DP_DTI(&dth->dth_xid), rc);
+		dth->dth_sync = 1;
+	}
+
+	if (dth->dth_sync) {
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid,
+				cont->sc_uuid, &dth->dth_dte, 1,
+				cont_hdl->sch_pool->spc_map_version);
+		if (rc != 0) {
+			D_ERROR(DF_UUID": Fail to sync commit DTX "DF_DTI
+				": rc = %d\n", DP_UUID(cont->sc_uuid),
+				DP_DTI(&dth->dth_xid), rc);
+			D_GOTO(fail, result = rc);
+		}
+	}
+
+fail:
+	if (result < 0 && dth->dth_leader)
+		dtx_abort(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+			  &dth->dth_dte, 1,
+			  cont_hdl->sch_pool->spc_map_version);
+out:
+	D_DEBUG(DB_TRACE,
+		"Stop the DTX "DF_DTI" ver %u, dkey %llu, intent %s, "
+		"%s, %s, %s: rc = %d\n",
+		DP_DTI(&dth->dth_xid), dth->dth_ver,
+		(unsigned long long)dth->dth_dkey_hash,
+		dth->dth_intent == DAOS_INTENT_PUNCH ? "Punch" : "Update",
+		dth->dth_sync ? "sync" : "async",
+		dth->dth_non_rep ? "non-replicated" : "replicated",
+		dth->dth_leader ? "leader" : "non-leader", result);
+
+	if (dth->dth_leader && dth->dth_dti_cos != NULL)
+		D_FREE(dth->dth_dti_cos);
+	D_FREE_PTR(dth);
+
+	return result > 0 ? 0 : result;
+}
+
```

Comment:
please add comment for this function, either in .c or .h

Reply:
OK